testtime/testtime.go

```diff
@@ -0,0 +1,201 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package testtime provides a mocked version of time.Timer for use in tests.
+package testtime
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/timeutil"
+)
+
+// TestTimer is a mocked version of time.Timer for which the passage of time or
+// the direct expiration of the timer is controlled manually.
+//
+// TestTimer implements timeutil.Timer.
+//
+// TestTimer also provides methods to introspect whether the timer is active or
+// how many times it has fired.
+type TestTimer struct {
+	lock       sync.Mutex
+	currTime   time.Time
+	expiration time.Time
+	active     bool
+	fireCount  int
+	callback   func()
+	c          chan time.Time
+}
+
+var _ timeutil.Timer = (*TestTimer)(nil)
+
+// AfterFunc waits for the timer to fire and then calls f in its own goroutine.
+// It returns a Timer that can be used to cancel the call using its Stop method.
+// The returned Timer's C field is not used and will be nil.
+//
+// AfterFunc returns a TestTimer which simulates the behavior of a timer which
+// was created via time.AfterFunc.
+//
+// See here for more details: https://pkg.go.dev/time#AfterFunc
+func AfterFunc(d time.Duration, f func()) *TestTimer {
+	currTime := time.Now()
+	return &TestTimer{
+		currTime:   currTime,
+		expiration: currTime.Add(d),
+		active:     true,
+		callback:   f,
+	}
+}
```

Comment:
with this you should be able to drop the code in run-checks
```suggestion
func AfterFunc(d time.Duration, f func()) *TestTimer {
	osutil.MustBeTestBinary("testtime timers cannot be used outside of unit test code")
	currTime := time.Now()
	return &TestTimer{
		currTime:   currTime,
		expiration: currTime.Add(d),
		active:     true,
		callback:   f,
	}
}
```

Reply:
I'm not sure if this works... If I make this change, plus the ones you suggest, and then run `go test` from inside `client/`, the test just hangs:
```diff
diff --git a/client/client.go b/client/client.go
index d9a936549c..ecb84b83e9 100644
--- a/client/client.go
+++ b/client/client.go
@@ -38,6 +38,7 @@ import (
 	"github.com/snapcore/snapd/features"
 	"github.com/snapcore/snapd/httputil"
 	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/testtime"
 )
 
 func unixDialer(socketPath string) func(string, string) (net.Conn, error) {
@@ -391,7 +392,7 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 		}
 		retry := time.NewTicker(opts.Retry)
 		defer retry.Stop()
-		timeout := time.NewTimer(opts.Timeout)
+		timeout := testtime.NewTimer(opts.Timeout)
 		defer timeout.Stop()
 
 		for {
@@ -409,7 +410,7 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 			select {
 			case <-retry.C:
 				continue
-			case <-timeout.C:
+			case <-timeout.ExpiredC():
 			}
 			break
 		}
```

Reply:
I think I'd prefer to leave the static check in place as well as the `osutil.MustBeTestBinary()` call, since the former can catch usage of `testtime.{AfterFunc,NewTimer}` even when the user is a package which is not imported elsewhere in the codebase.